### ISSUE TITLE
results for R code 4.53 is way off because of R code 4.39

### ISSUE THREAD

**Issue by RaymondBalise** [rmcelreath/rethinking#120]
If somebody runs the code in block 4.39 it breaks the code that uses the link function in the rest of the chapter.  

For example R code 4.53 gives:
> str(mu)
 num [1:1000, 1:352] 42.9 41.9 42.4 41.5 42.9 ...

... instead of the correct answers:
> str(mu)
 num [1:1000, 1:352] 157 157 157 157 157 ...

Wowza was that ever rough to figure out.  I don't know if this issue belongs in the errata but you want to mention it somewhere because the clue in the last sentence of the overthinking block at the top of page 97 didn't save me.

**Comment by ogletrees** [comment]
I also ran into this and can say that if you do not run block 4.39 everything comes out as expected in block 4.53. Glad this was here as I couldn't figure out what as going wrong. Thanks RaymondBalise! 

My understanding is that the _Overthinking_ boxes are for detail not essential for moving through the book. When using the text file of code these blocks are not indicated as _Overthinking_ sections so it is easy to keep running and get this unfortunate result (that's what I was doing).

**Comment by rplzzz** [comment]
**tl;dr:** Block 4.39 replaces `m4.3` with a model object that may not be usable with some of the package's analysis functions.

Code block 4.39 replaces `m4.3` with a version of the model that was run with the linear model written inline in the likelihood definition, like so.
```
...
    height ~ dnorm( a + b*weight, sigma)
...
```
The `map` object returned by this version of the call is very similar to the original, but there is one key difference.  Here's the original from block 4.3a (which I've called `m4.3a`):
```
> str(m4.3a)
Formal class 'map' [package "rethinking"] with 10 slots
  ..@ call          : language map(flist = alist(height ~ dnorm(mu, sigma), mu <- a + b * weight, a ~ dnorm(156, 100), b ~ dnorm(0, 10), sigma ~ dunif(0, 50)),      data = d2)

...

  ..@ links         :List of 1
  .. ..$ :List of 2
  .. .. ..$ : chr "mu"
  .. .. ..$ : chr "a + b * weight"
```
And here's the corresponding version from block 4.39:
```
> str(m4.3b)
Formal class 'map' [package "rethinking"] with 10 slots
  ..@ call          : language map(flist = alist(height ~ dnorm(a + b * weight, sigma), a ~ dnorm(156, 100), b ~ dnorm(0, 10), sigma ~ dunif(0, 50)), data = d2)

...

  ..@ links         : list()
```
This `links` slot is used in in the `link` function:
https://github.com/rmcelreath/rethinking/blob/a309712d904d1db7af1e08a76c521ab994006fd5/R/z_link-map.r#L35-L42
Having set `f_do_lm` to `FALSE`, the code will gamely attempt to guess the expression for `mu`.  Without going into the gory details, it winds up with this:
```
> as.character(m4.3b@formula[[1]][[3]][[2]]) -> form
> form
[1] "+"          "a"          "b * weight"
```
which _looks_ like it ought to be right, but for some reason parsing and evaluating that character string leads to something that only applies the "b*weight" part.  For example:
```
> a <- 2
> b <- 1
> weight <- 100    # a + b*weight should be 102
> eval(parse(text=form))
[1] 100
```
So, instead of getting a matrix of `mu` values, we get a bunch of `b*weight` values.

Finally, I'll call your attention to this text following block 4.39:
> Some of the helper functions you'll use later depend upon the explicit linear model however, so I recommend using it.

This appears to be an example of one such case.  Probably block 4.39 should have assigned its result to a different variable, such as `m4.3alt`, so as to avoid messing up the real `m4.3`.  I guess this could be updated in the downloadable code blocks on the website, but of course there's no fixing the printed version in the book (at least until another edition comes out).

**Comment by rplzzz** [comment]
Ha! I figured out why the expression isn't parsing correctly.  The problem is that this:
```
expression(+ a, b*weight)
```
doesn't evaluate to the intended expression:
```
`+`(a, b*weight)
```
I'm actually a little surprised it parsed at all.  R doesn't have a comma operator, so I'm not sure what `+a, b*weight` is supposed to be.

Anyhow, the fix seems easy.  The `as.character` call on line 58 seems to be unnecessary:
```
> m4.3b@formula[[1]][[3]][[2]] -> form2
> eval(form2)                               # a==2, b==1, and weight==100
[1] 102
```
So, removing it should cause this model, at least, to work properly.

**Comment by dxander** [comment]
in R code 4.53 and 4.54 simply change m4.3 to mN
a new model was calculated but the old one was used that caused the error.   After changing to model mN all graphs and numbers come out correct.

# R code 4.53
```{r   code4.53}
mu <- link( mN)
str(mu)
```
# R code 4.54
```{r code4.54}

# define sequence of weights to compute predictions for 
# these values will be on the horizontal axis
weight.seq <- seq( from = 25 , to = 70,  by = 1 )

# use link to compute mu
# for each sample from posterior
# and for each weight in weight.seq
mu <- link( mN , data = data.frame(weight = weight.seq) )
str(mu)
```

**Comment by dxander** [comment]
mN let's you choose how many data points you want to include ... the code starts with 10 but then you are supposed to increase it to the full data set to see the effect.   
## Samples from the quadratic approximate posterior distribution for the height/weight model, m4.3, with increasing amounts of data.  In each plot,
20 lines sampled form the posterior distribution, showing the uncertainty in the regression relationship.  

```{r code4.48}
N <- 352  # N = 10, 50, 150, 352
dN <- d2[ 1:N , ]
mN <- map(
     alist(
       height ~ dnorm( mu, sigma ) ,
       mu <- a + b * weight ,
       a ~ dnorm( 178 , 100 ) ,
       b ~ dnorm( 0 , 10 ) ,
       sigma ~ dunif( 0 , 50 )
     ) , data = dN )
```

**Comment by billmeade** [comment]
Dxander you rock! I was trying to debug Figure 4.7 & 4.8 in the raw R functions (abline etc.), and then in m4.3 vs m4.4 (which does not plot the way m4.3 plots). Thank you for posting the mN finding!
